### PR TITLE
2177 Add max_roi_size to RandSpatialCrop

### LINKED ISSUE #2177: Add max_roi_size support in `RandSpatialCrop`
**Is your feature request related to a problem? Please describe.**
Currently, when we do spatial crop, we can select a random size in (roi_size, image_size) where roi_size is the min_size. So we also need to support a max roi_size.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #2177 .

### Description
This PR added support to limit the max random crop size in `RandSpatialCrop`.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by rijobro** [comment]
Do you think it would look nicer if, instead of adding `max_roi_size`, we modified `roi_size` to also take `Sequence[Sequence[int]]` i.e., `roi_size=((10,10,10), (20, 20, 20))`? It's just a shame that we have `roi_size` and `max_roi_size`, where `roi_size` actually means `min_roi_size`.

File: monai/transforms/croppad/array.py [hunk 1]
```diff
@@ -342,9 +351,13 @@ class RandSpatialCropSamples(Randomizable):
     It will return a list of cropped images.
 
     Args:
-        roi_size: if `random_size` is True, the spatial size of the minimum crop region.
-            if `random_size` is False, specify the expected ROI size to crop. e.g. [224, 224, 128]
+        roi_size: if `random_size` is True, it specifies the minimum crop region.
+            if `random_size` is False, it specifies the expected ROI size to crop. e.g. [224, 224, 128]
+            If its components have non-positive values, the corresponding size of input image will be used.
         num_samples: number of samples (crop regions) to take in the returned list.
+        max_roi_size: if `random_size` is True and `roi_size` specified the min crop region size, `max_roi_size`
```
Comment (yiheng-wang-nv) [inline monai/transforms/croppad/array.py:358]: `roi_size` specifies

File: monai/transforms/croppad/array.py [hunk 2]
```diff
@@ -298,15 +298,23 @@ class RandSpatialCrop(Randomizable):
         roi_size: if `random_size` is True, it specifies the minimum crop region.
             if `random_size` is False, it specifies the expected ROI size to crop. e.g. [224, 224, 128]
             If its components have non-positive values, the corresponding size of input image will be used.
+        max_roi_size: if `random_size` is True and `roi_size` specified the min crop region size, `max_roi_size`
+            can specify the max crop region size. if None, default to the input image size.
```
Comment (yiheng-wang-nv) [inline monai/transforms/croppad/array.py:302]: defaults to

File: monai/transforms/croppad/array.py [hunk 3]
```diff
@@ -342,9 +351,13 @@ class RandSpatialCropSamples(Randomizable):
...
 
     Args:
-        roi_size: if `random_size` is True, the spatial size of the minimum crop region.
-            if `random_size` is False, specify the expected ROI size to crop. e.g. [224, 224, 128]
+        roi_size: if `random_size` is True, it specifies the minimum crop region.
+            if `random_size` is False, it specifies the expected ROI size to crop. e.g. [224, 224, 128]
+            If its components have non-positive values, the corresponding size of input image will be used.
         num_samples: number of samples (crop regions) to take in the returned list.
+        max_roi_size: if `random_size` is True and `roi_size` specified the min crop region size, `max_roi_size`
+            can specify the max crop region size. if None, default to the input image size.
```
Comment (yiheng-wang-nv) [inline monai/transforms/croppad/array.py:359]: defaults to

File: monai/transforms/croppad/dictionary.py [hunk 4]
```diff
@@ -410,6 +410,9 @@ class RandSpatialCropd(Randomizable, MapTransform, InvertibleTransform):
         roi_size: if `random_size` is True, it specifies the minimum crop region.
             if `random_size` is False, it specifies the expected ROI size to crop. e.g. [224, 224, 128]
             If its components have non-positive values, the corresponding size of input image will be used.
+        max_roi_size: if `random_size` is True and `roi_size` specified the min crop region size, `max_roi_size`
+            can specify the max crop region size. if None, default to the input image size.
```
Comment (yiheng-wang-nv) [inline monai/transforms/croppad/dictionary.py:414]: defaults to

File: monai/transforms/croppad/dictionary.py [hunk 5]
```diff
@@ -410,6 +410,9 @@ class RandSpatialCropd(Randomizable, MapTransform, InvertibleTransform):
         roi_size: if `random_size` is True, it specifies the minimum crop region.
             if `random_size` is False, it specifies the expected ROI size to crop. e.g. [224, 224, 128]
             If its components have non-positive values, the corresponding size of input image will be used.
+        max_roi_size: if `random_size` is True and `roi_size` specified the min crop region size, `max_roi_size`
```
Comment (yiheng-wang-nv) [inline monai/transforms/croppad/dictionary.py:413]: specifies

File: monai/transforms/croppad/dictionary.py [hunk 6]
```diff
@@ -499,9 +505,13 @@ class RandSpatialCropSamplesd(Randomizable, MapTransform):
...
         keys: keys of the corresponding items to be transformed.
             See also: monai.transforms.MapTransform
-        roi_size: if `random_size` is True, the spatial size of the minimum crop region.
-            if `random_size` is False, specify the expected ROI size to crop. e.g. [224, 224, 128]
+        roi_size: if `random_size` is True, it specifies the minimum crop region.
+            if `random_size` is False, it specifies the expected ROI size to crop. e.g. [224, 224, 128]
+            If its components have non-positive values, the corresponding size of input image will be used.
         num_samples: number of samples (crop regions) to take in the returned list.
+        max_roi_size: if `random_size` is True and `roi_size` specified the min crop region size, `max_roi_size`
+            can specify the max crop region size. if None, default to the input image size.
```
Comment (yiheng-wang-nv) [inline monai/transforms/croppad/dictionary.py:513]: defaults to

**Comment by Nic-Ma** [comment]
Hi @rijobro ,

Thanks for your suggestion, but the problem is that:
1. `roi_size` only acts as "min" size when `random_size=True`, if False, it's the fixed roi size to crop.
2. We also support to set **only 1 value for all the spatial dims** in `roi_size`, for example: `roi_size=128`, if we put `max_roi_size` input `roi_size` as you suggested, when crop a 2D image, what `roi_size=(128, 192)` means? It can be roi=128 and max_roi_size=192, or real roi_size=(128, 192), right?

What do you think?

Thanks.

File: monai/transforms/croppad/dictionary.py [hunk 7]
```diff
@@ -434,7 +439,8 @@ def __init__(
     def randomize(self, img_size: Sequence[int]) -> None:
         self._size = fall_back_tuple(self.roi_size, img_size)
         if self.random_size:
-            self._size = [self.R.randint(low=self._size[i], high=img_size[i] + 1) for i in range(len(img_size))]
+            max_size = img_size if self.max_roi_size is None else fall_back_tuple(self.max_roi_size, img_size)
+            self._size = [self.R.randint(low=self._size[i], high=max_size[i] + 1) for i in range(len(img_size))]
```
Comment (yiheng-wang-nv) [inline monai/transforms/croppad/dictionary.py:443]: Errors will happen if `low > high`, thus I suggest add a comparison between `roi_size` and `max_roi_size` in `__init__`, and could you please add include one more test case to cover this situation.

**Review by yiheng-wang-nv — COMMENTED** [review]
Thanks Nic, I added some comments for this PR.

**Comment by rijobro** [comment]
@Nic-Ma Ok makes sense. I suppose we can't change the behaviour of `roi_size` either, because of backwards compatibility. If we were designing it now, I would have `roi_size`, `min_roi_size` and `max_roi_size` where you can specify the first one or the latter two. Ah well, looks good! 

I'll approve but agree with @yiheng-wang-nv 's suggested changes.

**Review by rijobro — APPROVED** [review]

**Comment by Nic-Ma** [comment]
Hi @rijobro and @yiheng-wang-nv ,

Thanks for your review, I will try to update the PR to address all the comments.

File: monai/transforms/croppad/dictionary.py (hunk 7 above)
Comment (Nic-Ma) [reply monai/transforms/croppad/dictionary.py:445]: added check for min / max ROI size.

**Pull request merged**